src/shared/firewall-util-nft.c

```diff
@@ -1071,3 +1135,222 @@ int fw_nftables_add_local_dnat(
         /* table created anew; previous address already gone */
         return fw_nftables_add_local_dnat_internal(ctx, add, af, protocol, local_port, remote, remote_port, NULL);
 }
+
+static const char *const nfproto_table[] = {
+        [NFPROTO_ARP] = "arp",
+        [NFPROTO_BRIDGE] = "bridge",
+        [NFPROTO_INET] = "inet",
+        [NFPROTO_IPV4] = "ip",
+        [NFPROTO_IPV6] = "ip6",
+        [NFPROTO_NETDEV] = "netdev",
+};
+
+DEFINE_STRING_TABLE_LOOKUP(nfproto, int);
+
+#define NFT_SET_MSGS 3
+
+static int nft_set_element_op(bool add, const NFTSetContext *nft_set_context, void *element, size_t element_size) {
+        _cleanup_(sd_netlink_unrefp) sd_netlink *nfnl = NULL;
+        sd_netlink_message *transaction[NFT_SET_MSGS] = {};
+        _cleanup_free_ uint32_t *serial = NULL;
+        size_t tsize;
+        int r, nfproto;
+        const char *table, *set;
+
+        assert(nft_set_context);
+        nfproto = nft_set_context->nfproto;
+        table = nft_set_context->table;
+        assert(table);
+        set = nft_set_context->set;
+        assert(set);
+        assert(element);
+
+        r = sd_nfnl_socket_open(&nfnl);
+        if (r < 0)
+                return r;
+
+        r = sd_nfnl_message_batch_begin(nfnl, &transaction[0]);
+        if (r < 0)
+                return r;
+        tsize = 1;
+
+        if (add)
+                r = sd_nfnl_nft_message_new_setelems_begin(nfnl, &transaction[tsize], nfproto, table, set);
+        else
+                r = sd_nfnl_nft_message_del_setelems_begin(nfnl, &transaction[tsize], nfproto, table, set);
+        if (r < 0)
+                goto out_unref;
+
+        r = sd_nfnl_nft_message_add_setelem(transaction[tsize], 0, element, element_size, NULL, 0);
+        if (r < 0)
+                return r;
+
+        r = sd_nfnl_nft_message_add_setelem_end(transaction[tsize]);
+        if (r < 0)
+                return r;
+        ++tsize;
+        assert(tsize < ELEMENTSOF(transaction));
+        r = sd_nfnl_message_batch_end(nfnl, &transaction[tsize]);
+        if (r < 0)
+                return r;
+
+        ++tsize;
```

Comment:
Why is `tsize` incremented twice?

Reply:
See similar uses in for example [above line 1091](https://github.com/systemd/systemd/blob/9c3e2a2a97d41f3a758c5815421122fc3e3dd3fb/src/shared/firewall-util-nft.c#L1091), `tsize` is used to keep track of used space in the array and `assert()`s guard that the array size isn't exceeded at any point.